### PR TITLE
Encrypt as module: solution for deprecated Mcrypt in PHP 7.1

### LINKED ISSUE #14: PHP 7.1: Deprecated Mcrypt
https://github.com/kohana/core/issues/686
https://github.com/yclas/yclas/issues/2176

### PR DESCRIPTION
**Pull request by svenbw**
Should fix #14.
Encrypt class is moved to a module, there are two options: Original Mcrypt and a replacement based on openssl. It is possible to add additional encryption engines without breaking the old ones.

### CODE REVIEW

**Comment by neo22s** [comment]
damn! this is great job!! please let me review carefully, thanks a lot

**Comment by svenbw** [comment]
Thanks. I needed to take some compromises to keep things backward compatible.

File: modules/encrypt/classes/Kohana/Encrypt.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+			{
+				// No default encryption key is provided!
+				throw new Kohana_Exception('No encryption key is defined in the encryption configuration group: :group',
+					array(':group' => $name));
+			}
+
+			if ( ! isset($config['mode']))
+			{
+				// Add the default mode
+				$config['mode'] = MCRYPT_MODE_NOFB;
```
Comment (piotrgolasz) [inline modules/encrypt/classes/Kohana/Encrypt.php]: Shouldn't the default mode be changed to openssl?

File: modules/encrypt/classes/Kohana/Encrypt.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+			}
+
+			if ( ! isset($config['cipher']))
+			{
+				// Add the default cipher
+				$config['cipher'] = MCRYPT_RIJNDAEL_128;
```
Comment (piotrgolasz) [inline modules/encrypt/classes/Kohana/Encrypt.php]: shouldn't the cipher type be changed to aes-256-ctr (or aes-256-cbc) for that matter?

File: modules/encrypt/classes/Kohana/Encrypt.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	 *
+	 * @param   string  $key_config    encryption key or config array
+	 * @param   string  $mode          encryption mode
+	 * @param   string  $cipher        encryption cipher
+	 */
+	public function __construct($key_config, $mode = NULL, $cipher = NULL)
+	{
+		if (is_string($key_config))
+		{
+			$this->_engine = new Encrypt_Engine_Mcrypt($key_config, $mode, $cipher);
```
Comment (piotrgolasz) [inline modules/encrypt/classes/Kohana/Encrypt.php]: The default should be Openssl imho.

File: modules/encrypt/classes/Kohana/Encrypt/Engine/Mcrypt.php [hunk 4]
```diff
@@ -154,11 +64,8 @@ public function __construct($key, $mode, $cipher)
 	 * @param   string  $data   data to be encrypted
 	 * @return  string
 	 */
-	public function encode($data)
+	public function encrypt($data, $iv)
```
Comment (piotrgolasz) [inline modules/encrypt/classes/Kohana/Encrypt/Engine/Mcrypt.php]: We should always assume that a programmer programming is a monkey, with no knowledge of encryption. I would therefore suggest not including the IV in encrypt() function, but instead automatically generate the most random IV.

File: modules/encrypt/config/encrypt.php [hunk 5]
```diff
@@ -3,6 +3,7 @@
 return array(
 
 	'default' => array(
+		'type'       => 'mcrypt',
```
Comment (piotrgolasz) [inline modules/encrypt/config/encrypt.php]: Use openssl as default, pretty please?

File: modules/encrypt/classes/Kohana/Encrypt.php (hunk 1 above)
Comment (svenbw) [inline modules/encrypt/classes/Kohana/Encrypt.php]: I wanted to keep the code compatible with the original implementation: This assures that updating the Kohana system doesn't break old code or starts encrypting stuff using another method.

File: modules/encrypt/classes/Kohana/Encrypt.php (hunk 3 above)
Comment (svenbw) [inline modules/encrypt/classes/Kohana/Encrypt.php]: Again... compatibility

File: modules/encrypt/config/encrypt.php (hunk 5 above)
Comment (svenbw) [inline modules/encrypt/config/encrypt.php]: I think this van be changed to the new default, existing installations have an existing config.

**Comment by svenbw** [comment]
@piotrgolasz you're seem to know what you're doing and provide solid feedback. Do you think these changes make the modifications acceptable?
Note that an encryption type is now required, using this module in existing configurations will require a change in the code (pass a configuration array).
I still need to fix the iv thing.

**Comment by neo22s** [comment]
@svenbw we should do something so the code doesnt need to change... maybe a hack.. like hardcoded config? the rest looks great to me. I haven't tested it need to be hones, only read the code.

**Comment by piotrgolasz** [comment]
@svenbw https://github.com/piotrgolasz/kohana-encrypt I created my own solution for personal use. On the other topic you could create something like legacy_decrypt() where it would decrypt with mcrypt and encrypt with openssl for compatibility

**Comment by svenbw** [comment]
@piotrgolasz as @neo22s mentioned it's preferable that the code remains compatible the existing code.
I see your point, but for the time being I think we should remain compatible with the existing implementation.

**Comment by svenbw** [comment]
PR on development branch under #39, this one is on the master. Following the development flow (#30) this can be closed.

**Pull request closed without merging**